### PR TITLE
df: fix output if input path is device name

### LINKED ISSUE #3246: `df`: df doesn't check if the input path is a filesystem
For example when we pass `/dev/sda3` to `df`, it should print details of `/dev/sda3` instead of considering it as mount point/filesystem path. Below are some example outputs:
gnu df
```shell
$ df --out=source,target /dev/sda3 .
Filesystem     Mounted on
/dev/sda3      /root
/dev/sda3      /root
```
uutils df
```shell
../../../target/release/df --out=source,target /dev/sda3 .
Filesystem       Mounted on
udev             /dev
/dev/sda3      /root
```

### PR DESCRIPTION
**Pull request by cakebaker**
Currently, when providing a device name as a path, you get some output like:
```
$ ./target/debug/coreutils df /dev/sda2 --out=source,target
Filesystem     Mounted on
dev            /dev
```
With this PR, the output is now the same as the output from GNU df:
```
$ ./target/debug/coreutils df /dev/sda2 --out=source,target
Filesystem     Mounted on
/dev/sda2      /boot
```
Fixes #3246

### CODE REVIEW

File: src/uu/df/src/filesystem.rs [hunk 1]
```diff
@@ -67,6 +67,16 @@ where
...
+
+    let matches: Vec<&MountInfo> = mounts
+        .iter()
+        .filter(|mi| mi.dev_name.eq(&path.to_string_lossy()))
+        .collect();
+
+    if !matches.is_empty() {
+        return Some(matches[0]);
+    }
+
```
Comment (anastygnome) [inline src/uu/df/src/filesystem.rs:79]: why not use something like the following instead ? 

```rust 
    let matches: Option<&MountInfo> = mounts
        .iter()
        .filter(|mi| mi.dev_name.eq(&path.to_string_lossy()))
        .next();

    if !matches.is_none(){
        return matches;
    }
```

File: src/uu/df/src/filesystem.rs [hunk 2]
```diff
@@ -67,6 +67,16 @@ where
...
     let matches = mounts.iter().filter(|mi| path.starts_with(&mi.mount_dir));
     matches.max_by_key(|mi| mi.mount_dir.len())
 }
```
Comment (anastygnome) [inline src/uu/df/src/filesystem.rs:82]: Why not do something like this instead ? 

```rust
    let matches: Option<&MountInfo> = mounts
        .iter()
        .filter(|mi| mi.dev_name.eq(&path.to_string_lossy()))
        .next(); // If the argument is a file system, treat it as such.

    return matches.or(mounts
        .iter()
        .filter(|mi| path.starts_with(&mi.mount_dir))
        .max_by_key(|mi| mi.mount_dir.len()));
}
```

Comment (anastygnome) [inline src/uu/df/src/filesystem.rs:82]: Why not do something like this instead ? 

```rust
    let matches: Option<&MountInfo> = mounts
        .iter()
        .find(|mi| mi.dev_name.eq(&path.to_string_lossy()));

    matches.or_else(|| {
        mounts
            .iter()
            .filter(|mi| path.starts_with(&mi.mount_dir))
            .max_by_key(|mi| mi.mount_dir.len())
    })
}
```

**Review by anastygnome — CHANGES_REQUESTED** [review]
Nice, but could be made way faster by using the methods from iterators. 
find takes the first match,  and the or_else method allows to keep the variable so that the overall code is faster :)

File: src/uu/df/src/filesystem.rs [hunk 3]
```diff
@@ -67,6 +67,16 @@ where
     } else {
         path.as_ref().to_path_buf()
     };
+
```
Comment (anastygnome) [inline src/uu/df/src/filesystem.rs:70]: Why not do something like this instead ? 

```rust
    let matches: Option<&MountInfo> = mounts
        .iter()
        .find(|mi| mi.dev_name.eq(&path.to_string_lossy()));

    matches.or_else(|| {
        mounts
            .iter()
            .filter(|mi| path.starts_with(&mi.mount_dir))
            .max_by_key(|mi| mi.mount_dir.len())
    })
}
```

**Review by anastygnome — CHANGES_REQUESTED** [review]
Good, but there's room for improvement : 

find returns the first match (so index 0) or None, which we can then catch using the or_else clause. 

This allows for further compile time optimisations ;)

File: src/uu/df/src/filesystem.rs (hunk 3 above)
Comment (anastygnome) [reply src/uu/df/src/filesystem.rs:70]: maybe we should use the stater module for it too while we are at it !

Comment (cakebaker) [reply src/uu/df/src/filesystem.rs:70]: Thanks for your suggestion. This is indeed a much cleaner solution and I switched to it in the recent push.

**Pull request merged**